File: kafka/src/main/java/com/example/kafka/producer/MyPartitioner.java
package com.example.kafka.producer;

import org.apache.kafka.clients.producer.Partitioner;
import org.apache.kafka.common.Cluster;

import java.util.Map;

/**
 @author yulshi
 @create 2020/03/17 22:53
 */
public class MyPartitioner implements Partitioner {
  @Override
  public int partition(String topic, Object key, byte[] keyBytes, Object value, byte[] valueBytes, Cluster cluster) {
    return value.hashCode() % 4;
  }
  @Override
  public void close() {}
  @Override
  public void configure(Map<String, ?> configs) {}
}

File: spark/sparkcore/src/main/scala/com/example/sparkcore/partition/MyPartitioner.scala
package com.example.sparkcore.partition

import org.apache.spark.Partitioner

/**
 * @author yulshi
 * @create 2020/03/11 21:26
 */
class MyPartitioner(partitions: Int) extends Partitioner {

  override def numPartitions: Int = partitions

  override def getPartition(key: Any): Int = 0
}
